core/src/main/scala/ai/h2o/sparkling/backend/H2ODataFrame.scala

```diff
@@ -54,21 +53,48 @@ private[backend] class ExternalBackendH2ODataFrame(val frame: H2OFrame, val requ
     }
   }
 
-  override val expectedTypes: Option[Array[VecType]] = {
+  override val expectedTypes: Array[VecType] = {
     // prepare expected type selected columns in the same order as are selected columns
     val javaClasses = selectedColumnIndices.map(indexToSupportedType(_).javaClass)
-    Option(ExternalH2OBackend.prepareExpectedTypes(javaClasses))
+    ConversionUtils.expectedTypesFromClasses(javaClasses)
   }
 
   override def compute(split: Partition, context: TaskContext): Iterator[InternalRow] = {
-    new H2ODataFrameIterator {
+    new H2OChunkIterator[InternalRow] {
       private val chnk = frame.chunks.find(_.index == split.index).head
-      override val reader: Reader = new ExternalBackendReader(frameKeyName, split.index, chnk.numberOfRows,
-        chnk.location, expectedTypes.get, selectedColumnIndices, h2oConf)
+      override val reader: Reader = new Reader(frameId, split.index, chnk.numberOfRows,
```

Comment:
question to think about: Is reader used straight after it' creation? if not, shouldn't it be lazy val? It seems that the constructor of `Reader` creates connection to the REST server (https://github.com/h2oai/sparkling-water/pull/1948/files#diff-1a7545e7f6937e88d3f808a2d607e875R39).

Reply:
Great catch!

Reader should be used whenever we run any action on the obtained dataframe ( after h2o frame -> spark frame conversion). So therefore, the reader should be lazy.

Thanks!